python/ray/data/util/torch_utils.py

```diff
@@ -0,0 +1,618 @@
+                ) from e
+            if unsqueeze:
+                t = t.unsqueeze(1)
+            feature_tensors.append(t)
+
+        if len(feature_tensors) > 1:
+            feature_tensor = torch.cat(feature_tensors, dim=1)
+        else:
+            feature_tensor = feature_tensors[0]
+        return feature_tensor
+
+    if multi_input:
+        if type(column_dtypes) not in [list, tuple]:
+            column_dtypes = [column_dtypes] * len(columns)
+        return [
+            get_tensor_for_columns(columns=subcolumns, dtype=dtype)
+            for subcolumns, dtype in zip(columns, column_dtypes)
+        ]
+    else:
+        return get_tensor_for_columns(columns=columns, dtype=column_dtypes)
+
+
+def convert_ndarray_to_torch_tensor(
+    ndarray: np.ndarray,
+    dtype: Optional[torch.dtype] = None,
+    device: Optional[Union[str, "torch.device"]] = None,
+    pin_memory: bool = False,
+) -> torch.Tensor:
+    """Convert a NumPy ndarray to a Torch Tensor.
+
+    Args:
+        ndarray: A NumPy ndarray that we wish to convert to a Torch Tensor.
+        dtype: A Torch dtype for the created tensor; if None, the dtype will be
+            inferred from the NumPy ndarray data.
+        device: The device on which the tensor(s) should be placed; if None, the Torch
+            tensor(s) will be constructed on the CPU.
+        pin_memory: Whether to pin the memory of the created tensors.
+
+    Returns:
+         A Torch Tensor.
+    """
+    ndarray = _unwrap_ndarray_object_type_if_needed(ndarray)
+
+    # Object dtype cannot be converted into PyTorch Tensor.
+    if ndarray.dtype.type is np.object_:
+        raise RuntimeError(
+            "Numpy array of object dtype cannot be converted to a Torch Tensor. This "
+            "may because the numpy array is a ragged tensor--it contains items of "
+            "different sizes. If using `iter_torch_batches()` API, you can pass in a "
+            "`collate_fn` argument to specify custom logic to convert the Numpy array "
+            "batch to a Torch tensor batch."
+        )
+
+    # The numpy array is not always writeable as it can come from the Ray object store.
+    # Numpy will throw a verbose warning here, which we suppress, as we don't write
+    # to the tensors. We also don't want to copy the array to avoid memory overhead.
+    # Original warning: https://github.com/pytorch/pytorch/blob/v1.13.0/
+    # torch/csrc/utils/tensor_numpy.cpp#L198-L206
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        result = torch.as_tensor(ndarray, dtype=dtype, device=device)
+
+    if pin_memory:
+        assert result.device.type == "cpu", (
+            "Pin memory is only supported for CPU tensors. "
+            f"Got device: {result.device} and pin_memory: {pin_memory}."
+        )
+        result = result.pin_memory()
+
+    return result
+
+
+def convert_ndarray_batch_to_torch_tensor_batch(
```

Comment:
I don't think we use this anywhere in Ray Data. I think we should remove this in a follow-up